aten/src/ATen/native/mkldnn/Conv.cpp

```diff
@@ -102,121 +103,46 @@ Tensor _mkldnn_conv2d(
 }
 
 at::Tensor mkldnn_convolution(
-    const at::Tensor& input, const at::Tensor& weight, const at::Tensor& bias,
-    IntArrayRef padding, IntArrayRef stride, IntArrayRef dilation, int64_t groups)
-{
+    const at::Tensor& input,
+    const at::Tensor& weight,
+    const at::Tensor& bias,
+    IntArrayRef padding,
+    IntArrayRef stride,
+    IntArrayRef dilation,
+    int64_t groups) {
+  Tensor mkldnn_input, mkldnn_weight, mkldnn_bias;
   if (input.is_mkldnn()) {
-    return _mkldnn_conv2d(input, weight, bias, padding, stride, dilation, groups);
-  }
-
-  auto output = at::empty(
-      conv_output_size(
-          input.sizes(), weight.sizes(), padding, stride, dilation),
-      input.options());
-
-  auto cpu_engine = CpuEngine::Instance().get_engine();
-
-  int32_t g = groups;
-
-  int32_t n = input.size(0);
-  int32_t ic = input.size(1);
-  int32_t ih = input.size(2);
-  int32_t iw = input.size(3);
-
-  int32_t oc = output.size(1);
-  int32_t oh = output.size(2);
-  int32_t ow = output.size(3);
-
-  int32_t kh = weight.size(2);
-  int32_t kw = weight.size(3);
-
-  int32_t sh = stride[0];
-  int32_t sw = stride[1];
-  int32_t ph = padding[0];
-  int32_t pw = padding[1];
-
-  auto data_t = memory::data_type::f32;
-  auto format_any = memory::format::any;
-  auto format_nchw = memory::format::nchw;
-  auto format_weight = (g!= 1) ? memory::format::goihw : memory::format::oihw;
-  auto format_x = memory::format::x;
-
-  memory::dims input_tz = {n, ic, ih, iw};
-  memory::dims weight_tz = (g!= 1) ? memory::dims{g, oc/g, ic/g, kh, kw} : memory::dims{oc, ic, kh, kw};
-  memory::dims bias_tz = {oc};
-  memory::dims output_tz = {n, oc, oh, ow};
-  memory::dims _stride = {sh, sw};
-  memory::dims _padding = {ph, pw};
-
-  auto input_md = memory::desc({input_tz}, data_t, format_any);
-  auto weight_md = memory::desc({weight_tz}, data_t, format_any);
-  auto bias_md = memory::desc({bias_tz}, data_t, format_any);
-  auto output_md = memory::desc({output_tz}, data_t, format_any);
-
-  std::shared_ptr<convolution_forward::desc> conv_forward_desc;
-  if (bias.defined()) {
-    conv_forward_desc.reset(new convolution_forward::desc(prop_kind::forward,
-      convolution_direct, input_md, weight_md, bias_md, output_md,
-      _stride, _padding, _padding, padding_kind::zero));
+    mkldnn_input = input;
+    AT_ASSERTM(weight.is_mkldnn(), "weight is supposed to be a mkldnn tenser");
+    mkldnn_weight = weight;
+    if (bias.defined()) {
+      AT_ASSERTM(bias.is_mkldnn(), "bias is supposed to be a mkldnn tensor");
+      mkldnn_bias = bias;
+    }
   } else {
-    conv_forward_desc.reset(new convolution_forward::desc(prop_kind::forward,
-      convolution_direct, input_md, weight_md, output_md,
-      _stride, _padding, _padding, padding_kind::zero));
+    mkldnn_input = at::native::mkldnn_reorder_conv2d_input(
```

Comment:
The extra same format reordering can be avoided if we check the current format and required format are actually the same.  The check was in original code path -   if (input_usr_memory.get_primitive_desc() != memory::primitive_desc(input_pd)) {
This explains why we have 10 extra reorderings, which could be optimized away with an appropriate check and buffer sharing between CPU tensor and MKL-DNN tensor.

Reply:
i think we don't need `mkldnn_reorder_conv2d_input` here. In case input is a cpu tensor, just new a `ideep::tensor` with `data_ptr()` of the cpu tensor will be fine. And the `itensor` will be sharing the same buffer with the cpu tensor, reorder(if necessary) is handled by `ideep` automatically.

Reply:
Agreed. @bddppq , our three have an internal discussion and think it is better to have the ideep to get the required format and reorder for it.  Mingfei has suggested the code below which convert from cpu tensor to ideep tensor without involving reordering (copying). This is our preferred approach.